Prefer putting the bookmark in the data directory

Konubinix opened the pull request:
Bookmark may be considered as data. Therefore It looks sensible to think that
some users may not want their bookmarks to pollute their configuration
directory.

Use the ~/.config path if it exists so as not to break backward compatibility.

Konubinix commented:
I opened a new issue instead of #1563. In #1563, I proposed to merge my master branch, which is not practical when wanting to continue the development while waiting for the branch to be merged.

Konubinix commented:
This pull request does what we decided in #1563 (at least, what I understood of it :-)). The datadir is used by default if the configdir does not exist. That way, new user have a more standard location of bookmarks while old user don't see any impact.

The-Compiler commented:
Could you please do the same for quickmarks too for consistency?

The old path is also still used in some other places:

- `misc/userscripts/qutedmenu` (I'm okay with hardcoding the new path there)
- `scripts/importer.py`
- `doc/qutebrowser.1.asciidoc` (for quickmarks, seems bookmarks aren't even in there...)

Finally, please don't use `@property`. It has the unfortunate consequence of hiding any exceptions happening in it with PyQt objects (for some reasons I can't remember, but AFAIK PyQt can't do anything about it). Better to just have an ordinary function.

Konubinix commented:
The scripts/importer.py does not appear to use the bookmark directory. It only prints the netscape bookmark on the standard output.

Konubinix commented:
Done :-)

The-Compiler commented:
Thanks! You're still using `@property` though :wink:

Konubinix commented:
Sorry, I missed that part ...

Konubinix commented:
Done ?

The-Compiler commented on qutebrowser/browser/urlmarks.py:
```diff
@@ -141,12 +141,21 @@ class QuickmarkManager(UrlMarkManager):
 
     def _init_savemanager(self, save_manager):
-        filename = os.path.join(standarddir.config(), 'quickmarks')
+        filename = self._quickmark_directory()
```
Still need to join with the filename here if I'm not mistaken.

The-Compiler commented on qutebrowser/browser/urlmarks.py:
```diff
@@ -141,12 +141,21 @@ class QuickmarkManager(UrlMarkManager):
         - removed gets emitted with the name as argument.
     """
 
+    def _quickmark_directory(self):
```
Please also add a small docstring to explain the situation.

The-Compiler commented on qutebrowser/browser/urlmarks.py:
```diff
@@ -243,18 +252,24 @@ class BookmarkManager(UrlMarkManager):
           second argument.
         - removed gets emitted with the URL as argument.
     """
+    def _bookmarks_directory(self):
```
Please add a docstring here too

The-Compiler commented on qutebrowser/browser/urlmarks.py:
```diff
@@ -141,12 +141,21 @@ class QuickmarkManager(UrlMarkManager):
+        else:
+            return standarddir.data()
+
+
```
flake8 (code style checker) is going to complain about two instead of one line here (between methods) :wink:

The-Compiler commented:
I added a few line comments, after that it's ready to test/merge I think. I'll try to get it into v0.7.0, but no promises yet :wink:

The-Compiler commented:
Oh, seems various [tests are failing](https://travis-ci.org/The-Compiler/qutebrowser/jobs/136651053) too. Can you please take a look?

lahwaacz commented on doc/qutebrowser.1.asciidoc:
```diff
@@ -130,7 +130,7 @@ show it.
 
 - '~/.config/qutebrowser/qutebrowser.conf': Main config file.
-- '~/.config/qutebrowser/quickmarks': Saved quickmarks.
+- '~/.local/share/qutebrowser/quickmarks': Saved quickmarks.
```
Can you change the order so that the `~/.config/` and `~/.local/share/` paths are not interleaved?

Konubinix commented:
Hmmm, it looks like some machine don't have a standard configdir. I added a check to use it only if available.

Konubinix commented:
I have the feeling that this time, the failing tests are not due to the patch itself. Am I right?

The-Compiler commented:
yep - OS X is failing on Travis currently, I'll try to take a look today.

lahwaacz commented on qutebrowser/browser/urlmarks.py:
```diff
@@ -244,17 +264,35 @@ class BookmarkManager(UrlMarkManager):
+        if inconfigdir is not None and os.path.isdir(inconfigdir):
+            return inconfigdir
+        else:
+            return indatadir
```
Keeping the return statements at the end of the method is useless and unreadable -- there should be only one branch deciding which path to use, not two. What about this?

```python
        config_dir = standarddir.config()
        if config_dir is not None:
            inconfigdir = os.path.join(config_dir, 'bookmarks')
            if os.path.isdir(inconfigdir):
                return inconfigdir
        return os.path.join(standarddir.data(), 'bookmarks')
```

On second thought, if `standarddir.config()` can return `None`, are we sure that `standarddir.data()` doesn't?

Konubinix replied:
Hi,
Thanks for the comment.
Jakub Klinkovský <notifications@github.com> writes:

[...]

> On second thought, if `standarddir.config()` can return `None`, are we sure that `standarddir.data()` doesn't?

The constructor of urlmark already checks the standarddir.data. May be
we can recheck here for sake of clarity or defer the check to the
constructor.

-- 
Konubinix
GPG Key    : 7439106A
Fingerprint: 5993 BE7A DA65 E2D9 06CE  5C36 75D2 3CED 7439 106A

lahwaacz replied:
Oh, that's even worse than I thought. Returning early from the parent constructor does not ensure that the object is in failed state or which features are disabled. As long as the object lives, it should be possible to call any of its methods - here `os.path.join` would raise `TypeError` if it gets `None` in the first argument. The check for `None` should be here and I think `_bookmarks_directory` should also just return `None` in that case.

lahwaacz commented on qutebrowser/browser/urlmarks.py:
```diff
@@ -141,12 +141,33 @@ class QuickmarkManager(UrlMarkManager):
+        data_dir = standarddir.data()
+        if config_dir is not None:
+            isinconfigdir = os.path.isfile(
+                os.path.join(config_dir, 'quickmarks')
```
Here you need to assemble the full file path to make the decision, return the directory name and later assemble the file path again, which is somewhat useless. Can't this method return the full path to the `quickmarks` file directly?

Konubinix replied:
Jakub Klinkovský <notifications@github.com> writes:

>> @@ -141,12 +141,33 @@ class QuickmarkManager(UrlMarkManager):
>>          - removed gets emitted with the name as argument.
>>      """
>>  
>> +    def _quickmark_directory(self):
>> +        """Return the directory containing the quickmarks.
>> +
>> +        For quickmarks are data, return by default the standard data
>> +        directory. In order not to break the compatibility with an older
>> +        behavior, if quickmarks are found in the standard configuration
>> +        directory, return this directory instead.
>> +        """
>> +        config_dir = standarddir.config()
>> +        data_dir = standarddir.data()
>> +        if config_dir is not None:
>> +            isinconfigdir = os.path.isfile(
>> +                os.path.join(config_dir, 'quickmarks')
>
> Here you need to assemble the full file path to make the decision, return the directory name and later assemble the file path again, which is somewhat useless. Can't this method return the full path to the `quickmarks` file directly?

I tried to fit as well as possible to the current structure of the
code. IIUC, the construction of LineParser expects a directory and a
file basename, then if I return a file in _quickmark_directory (that
would be renamed _quickmark_file :-)), I will need to split it in
_init_lineparser.

IMHO, both solutions are equally readable and understandable. Plus, for
they don't go in the performance critical path, readability matters most
here.

I don't mind providing _quickmark_file instead of _quickmark_directory,
but I seriously doubt it will be better.

-- 
Konubinix
GPG Key    : 7439106A
Fingerprint: 5993 BE7A DA65 E2D9 06CE  5C36 75D2 3CED 7439 106A

lahwaacz commented on qutebrowser/browser/urlmarks.py:
```diff
@@ -91,7 +91,7 @@ def __init__(self, parent=None):
         self._lineparser = None
 
-        if standarddir.config() is None:
+        if standarddir.data() is None:
```
This condition used to do what the `UrlMarkManager`'s docstring says:
```
        _lineparser: The LineParser used for the marks, or None
                     (when qutebrowser is started with -c '').
```
What happens now if you start `qutebrowser -c ""`? `_lineparser` would not be `None` and I guess there would be bad side effects.

Not to mention that the whole condition belongs to `_init_lineparser` and the constructor should check if `self._lineparser` is still `None` after calling `_init_lineparser`. Somebody should do some refactoring here... But as that code is not your fault, I apologize for bringing it up here. Please feel free to take care of this or leave it, I can create a separate issue/PR later.

The-Compiler commented:
Sorry for the silence from my side here - will try to take a look at your comments and review this later today or tomorrow.

The-Compiler commented:
I kind of lost track of this, sorry - @Konubinix, how are things looking like? I think the comments from @lahwaacz still apply, no?

Konubinix commented:
Florian Bruhin <notifications@github.com> writes:

> I kind of lost track of this, sorry

So did I :-). I still use my forked version...

> - @Konubinix, how are things looking like? I think the comments from
> @lahwaacz still apply, no?

I won't have time to look at this till end of august (holidays ...). I
will answer then.

-- 
Konubinix
GPG Key    : 7439106A
Fingerprint: 5993 BE7A DA65 E2D9 06CE  5C36 75D2 3CED 7439 106A

The-Compiler commented:
No worries - enjoy your holidays! :palm_tree:

Konubinix commented:
Hi!
I am looking at this issue again.

@lahwaacz, you're right, that would break the backward compatibility behavior I have set up.

I have the feeling that handling backward compatible change will make the code quite ugly and I am not sure it is worth it.

I think I will go with my own fork of qutebrowser for now.

Thank you for helping me investigate he issue :-)

The-Compiler commented:
Fair enough - I'll try to take this over some time (unless someone else does) as I think there should be some clean solution, but right now I'm quite busy with other work :wink:

The-Compiler commented:
I still haven't had the time to look at this, but as a quick update: The `-c` (and other `--*dir`) arguments are now removed, probably making this a bit more straightforward too.

Konubinix commented:
Florian Bruhin <notifications@github.com> writes:

> I still haven't had the time to look at this, but as a quick update: The `-c` (and other `--*dir`) arguments are now removed, probably making this a bit more straightforward too.

I just updated the patch to rebase it on master and take into account
the possibility of config.datadir() being None.

The tests are not finished yet.

-- 
Konubinix
GPG Key    : 7439106A
Fingerprint: 5993 BE7A DA65 E2D9 06CE  5C36 75D2 3CED 7439 106A

Konubinix commented:
... and just made it pip8 compliant

Konubinix commented:
If I understand correctly the results of the tests. There is no regression in the tests themselves, but the code has fewer coverage. The part that is concerned is the handling of the backward compatibilty in case there already exist some bookmarks/quickmarks in the old location.

Konubinix commented:
I am not against a little help to implement the associated test.

IIUC, I should add a scenario in the urlmarks.feature file to check that given I already have some urls, the behavior is to take them.

IMHO, the correct way to do this would make a scenario look like this
  Scenario: Use the bookmarks from the old bookmark directory if they exist
    Given I have data/urls_with_thirteen in {config_dir}/bookmarks/urls
    And I have a fresh instance
    When I open qute:bookmarks
    Then the page should contain the plaintext "thirteen"

I should also add the scenario to check
- {data_dir}/bookmarks/urls taken if {config_dir}/bookmarks does not exist
- the same, with quickmarks

I think I should change the step implementation to use {config_dir} and {data_dir}.

What is your opinion?

The-Compiler commented:
Glad to see you picked this up again, I didn't get to it so far :wink:

With the change I mentioned above, it's not possible for the data/config dir to be `None` anymore, so you can remove those checks again. Sorry if that wasn't clear!

Your observation is correct - all existing tests passed, but indeed it'd be nice to have new one.

I don't think you can easily test initialization stuff like that in a `.feature` file, I'd probably write a test in python in `test_invocations.py` for this.

There you could use the [tmpdir fixture](http://doc.pytest.org/en/latest/tmpdir.html) of pytest to set up the necessary directory structure, start qutebrowser with `['--basedir', str(tmpdir)]` or so, and then make sure you can load some quickmark/bookmark from there and the old file got deleted.

I hope this made sense? :wink:

Konubinix commented:
Florian Bruhin <notifications@github.com> writes:

> With the change I mentioned above, it's not possible for the data/config dir to be `None` anymore, so you can remove those checks again. Sorry if that wasn't clear!

Yes it was. I just already made the change.

[...]

> There you could use the [tmpdir fixture](http://doc.pytest.org/en/latest/tmpdir.html) of pytest to set up the necessary directory structure, start qutebrowser with `['--basedir', str(tmpdir)]` or so, and then make sure you can load some quickmark/bookmark from there and the old file got deleted.

I'll try that.

Thank you for the answer.
-- 
Konubinix
GPG Key    : 7439106A
Fingerprint: 5993 BE7A DA65 E2D9 06CE  5C36 75D2 3CED 7439 106A

Carpetsmoker commented:
Honestly, I find this entire PR dubious. I *configure* my bookmarks and quickmarks to a certain state. It seems like *configuration* to me.

More pragmatically, right now I can do a backup of my qutebrowser settings with `cp -r ~/.config/qutebrowser`. However, if this is changed this, I 1) have to *also* copy `~/.local/share/qutebrowser` *and* 2) remember to filter our the sensitive data, such as cookies.

More pragmatically number two: it's *either* a backwards-incompatibel change, *or* you're going to complicate the config file search path, both of which have great potential to cofuse eithers and are great things to do, IMHO.

The-Compiler commented:
I don't really have a strong opinion on this - but since this has been inactive since some time again, I'm closing this. Thanks @Konubinix for the PR anyways!

lahwaacz commented:
> More pragmatically, right now I can do a backup of my qutebrowser settings with `cp -r ~/.config/qutebrowser`. However, if this is changed this, I 1) have to *also* copy `~/.local/share/qutebrowser` *and* 2) remember to filter our the sensitive data, such as cookies.

That's pragmatic only under the assumption that people _don't_ consider bookmarks and quickmarks as sensitive data. At least for me bookmarks are as sensitive as the browsing history, which I definitely don't want to share publicly. I do want to share qutebrowser's configuration like `qutebrowser.conf` and `keys.conf`, so I already have to do some filtering.

Konubinix commented:
Florian Bruhin <notifications@github.com> writes:

> I don't really have a strong opinion on this - but since this has been inactive since some time again, I'm closing this. Thanks @Konubinix for the PR anyways!

You did well :-). I don't have time anymore to work on this. Plus, I
also think that the difference between configuration vs data is
subjective.

Thank you anyway for taking the time to investigate my PR :-).

-- 
Konubinix
GPG Key    : 7439106A
Fingerprint: 5993 BE7A DA65 E2D9 06CE  5C36 75D2 3CED 7439 106A
